be/src/storage/storage_engine.cpp

```diff
@@ -915,19 +915,41 @@ Status StorageEngine::_do_sweep(const std::string& scan_root, const time_t& loca
     return res;
 }
 
-void StorageEngine::start_delete_unused_rowset() {
-    std::lock_guard lock(_gc_mutex);
-    for (auto it = _unused_rowsets.begin(); it != _unused_rowsets.end();) {
-        if (it->second.use_count() != 1) {
-            ++it;
-        } else if (it->second->need_delete_file()) {
-            VLOG(3) << "start to remove rowset:" << it->second->rowset_id()
-                    << ", version:" << it->second->version().first << "-" << it->second->version().second;
-            Status status = it->second->remove();
-            VLOG(3) << "remove rowset:" << it->second->rowset_id() << " finished. status:" << status;
-            it = _unused_rowsets.erase(it);
+void StorageEngine::start_delete_unused_rowset(int32_t* interval) {
+    MonotonicStopWatch timer;
+    timer.start();
+    std::vector<RowsetSharedPtr> delete_rowsets;
+    int64_t batch_size = 4096;
+    delete_rowsets.reserve(batch_size);
+    {
+        std::lock_guard lock(_gc_mutex);
+        for (auto it = _unused_rowsets.begin(); it != _unused_rowsets.end();) {
+            if (it->second.use_count() != 1) {
+                ++it;
+            } else if (it->second->need_delete_file()) {
+                delete_rowsets.emplace_back(it->second);
+                it = _unused_rowsets.erase(it);
+            }
+            if (delete_rowsets.size() > batch_size) {
+                auto new_interval = *interval / (_unused_rowsets.size() + batch_size / batch_size);
+                *interval = new_interval > 0 ? new_interval : 1;
```

Comment:
And I think it is not a good way to change interval in this function.
you can return `number_removed_rowsests`, and let caller decide what will do next